plans/backup.pp

```diff
@@ -0,0 +1,66 @@
+# @summary Backup the core user settings for puppet infrastructure
+#
+# This plan can backup data as outlined at insert doc 
+# 
+plan peadm::backup (
+  # Standard
+  Peadm::SingleTargetSpec           $primary_host,
+  Optional[Peadm::SingleTargetSpec] $replica_host            = undef,
+
+  # Large
+  Optional[TargetSpec]              $compiler_hosts          = undef,
+
+  # Extra Large
+  Optional[Peadm::SingleTargetSpec] $primary_postgresql_host = undef,
+  Optional[Peadm::SingleTargetSpec] $replica_postgresql_host = undef,
+
+  # Which data to backup
+  Boolean                            $backup_orchestrator    = true,
+  Boolean                            $backup_rbac            = true,
+  Boolean                            $backup_activity        = true,
+  Boolean                            $backup_ca_ssl          = true,
+  Boolean                            $backup_puppetdb        = false,
+  Boolean                            $backup_classification  = true,
+  String                             $backup_directory       = '/tmp'
+){
+  # Create an array of the names of databases and whether they have to be backed up to use in a lambda later
+  $database_to_backup = [ $backup_orchestrator, $backup_activity, $backup_rbac, $backup_puppetdb]
+  $database_names     = [ 'pe-orchestrator' , 'pe-activity' , 'pe-rbac' , 'pe-puppetdb' ]
+
+  # Database backups should take place on the postgress server
+  if $primary_postgresql_host {
+    $database_backup_server = $primary_postgresql_host
+  } else {
+    $database_backup_server = $primary_host
+  }
+
+  peadm::assert_supported_bolt_version()
+
+  # Ensure input valid for a supported architecture
+  $arch = peadm::assert_supported_architecture(
+    $primary_host,
+    $replica_host,
+    $primary_postgresql_host,
+    $replica_postgresql_host,
+    $compiler_hosts,
+  )
+
+  if $backup_classification {
+    out::message('# Backing up classification')
+    run_task('peadm::backup_classification', $primary_host,
+    directory => $backup_directory,
+    )
+  }
+
+  if $backup_ca_ssl {
+    out::message('# Backing up ca and ssl certificates')
+    run_command("/opt/puppetlabs/bin/puppet-backup create --dir=${backup_directory} --scope=certs", $primary_host)
+  }
+
+  $database_to_backup.each |Integer $index, Boolean $value | {
+    if $value {
+    out::message("# Backing up database ${database_names[$index]}")
+    run_command("sudo -u pe-postgres /opt/puppetlabs/server/bin/pg_dump -Fc \"${database_names[$index]}\" -f \"${backup_directory}/${database_names[$index]}_$(date +%Y%m%d%S).bin\" || echo \"Failed to dump database ${database_names[$index]}\"" , $database_backup_server) # lint:ignore:140chars
```

Comment:
Two comments on this line.

**First:**

When this plan creates a backup, it should always create a single backup archive in a single location, regardless of which hosts the backup data comes from. This is a PE cluster backup, not a per-host backup.

Therefore, the host that this command runs on should probably always be the primary server. We can change that later if we want but just picking the place the backups go is the simplest thing to do for now.

The `pg_dump` command should be runnable from the primary server. It should still be able to connect to a different host to retrieve the backups, even when running from the primary. It'll be necessary to figure out the correct connection string to do it, and to pass a per-database username.

Note also that even when there is a separate PostgreSQL server, that PostgreSQL server only hosts PuppetDB data. The rest of the databases are always still served from the primary server.

**Second:**

```suggestion
    run_command("sudo -u pe-postgres /opt/puppetlabs/server/bin/pg_dump -Fc \"${database_names[$index]}\" -f \"${output_directory}/${database_names[$index]}_$(date +%Y%m%d%S).bin\" || echo \"Failed to dump database ${database_names[$index]}\"" , $database_backup_server) # lint:ignore:140chars
```